domain/resource/state/resource.go

```diff
@@ -1009,30 +1012,45 @@ ON CONFLICT(resource_uuid) DO UPDATE SET retrieved_by_type_id=excluded.retrieved
 // application associated with a resource.
 func (st *State) incrementCharmModifiedVersion(ctx context.Context, tx *sqlair.TX, resourceUUID coreresource.UUID) error {
 	resID := resourceIdentity{UUID: resourceUUID.String()}
+	getApplicationUUIDsStmt, err := st.Prepare(`
+SELECT &applicationUUID.*
+FROM   application_resource
+WHERE  resource_uuid = $resourceIdentity.uuid
+	`, applicationUUID{}, resID)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
 	updateCharmModifiedVersionStmt, err := st.Prepare(`
 UPDATE application
-SET    charm_modified_version = IFNULL(charm_modified_version ,0) + 1
-WHERE  uuid IN (
-    SELECT application_uuid
-    FROM   application_resource
-    WHERE  resource_uuid = $resourceIdentity.uuid
-)
-`, resID)
+SET    charm_modified_version = $charmModifiedVersion.charm_modified_version
+WHERE  uuid = $applicationUUID.application_uuid
+`, charmModifiedVersion{}, applicationUUID{})
 	if err != nil {
 		return errors.Capture(err)
 	}
 
-	var outcome sqlair.Outcome
-	err = tx.Query(ctx, updateCharmModifiedVersionStmt, resID).Get(&outcome)
-	if err != nil {
+	var appUUIDs []applicationUUID
+	err = tx.Query(ctx, getApplicationUUIDsStmt, resID).GetAll(&appUUIDs)
+	if errors.Is(err, sqlair.ErrNoRows) || len(appUUIDs) > 1 {
+		return errors.Errorf("updating charm modified version: expected 1 application for resource, got %d", len(appUUIDs))
+	} else if err != nil {
 		return errors.Errorf("updating charm modified version: %w", err)
 	}
```

Comment:
issue: The resource_uuid is the primary_key of the application_resource, you won't get multiple application_uuids from this query. Not sure why it's broken into two pieces.